### ISSUE TITLE
histogram_quantile: Outputs wrong values that match the upper bound of the largest histogram bucket

### ISSUE THREAD

**Issue by AlbertoEAF** [prometheus/prometheus#13020]
### What did you do?

Hello,

I am doing some monitoring in Grafana with Prometheus' `histogram_quantile()` and have been observing some strange outputs that are disturbing our latency monitoring for high quantiles, like `0.9999`.

---

From time to time `histogram_quantile()`  creates peaks where the reported latency is no longer what is expected (~5-200ms @ q0.9999, as confirmed with exact external load test tools), but jumps several orders of magnitude to exactly `60s` -- the largest bucket upper bound of the underlying Prometheus histogram.

This issue arises more frequently with higher percentiles, and is specially frequent and obvious when I used the limit `quantile=1` for this issue report:

![image](https://github.com/grafana/grafana/assets/3976753/e4ff115c-0c50-4652-80ed-881130d6e069)

---

Let's zoom in on one of those `60s` points:

![image](https://github.com/grafana/grafana/assets/3976753/eaf8c01e-5b52-48d8-bcc3-eb7274f3b5a6)

If I zoom in on that point alone, remove the `histogram_quantile()` operation and sort the buckets' `le` bounds in descending order we can see that the all the largest buckets have all the samples:
![image](https://github.com/grafana/grafana/assets/3976753/fe699fcb-4b22-4493-acc9-604f66909b92)

In fact, the largest bucket that doesn't contain 100% of the samples is found at `le=0.0168s=16.8ms`, right before `le=0.0224`, which would be the expected output:

![image](https://github.com/grafana/grafana/assets/3976753/09b7c9e7-d95a-4715-a5ee-8cdabdfdbbd7)

----

I'm going to keep this Grafana Explore link valid only to me here, in case you want to see the raw buckets' data near this timestamp if that's useful to reproduce ([private hashed link for future reference](https://grafana.monitoring.eu-01.xxxxxxxxxxxxxx/explore?left=%7B%22datasource%22:%22prometheus%22,%22queries%22:%5B%7B%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22prometheus%22%7D,%22editorMode%22:%22code%22,%22exemplar%22:true,%22expr%22:%22histogram_quantile%281,%20sum%20by%28le%29%20%28rate%28http_server_requests_seconds_bucket%7BprojectId%3D%5C%22influencer-model-dt%5C%22%7D%5B2m%5D%29%29%29%22,%22format%22:%22time_series%22,%22hide%22:true,%22instant%22:false,%22interval%22:%22%22,%22range%22:true,%22refId%22:%22histogram_quantile%281,%20histogram_buckets%29%22%7D,%7B%22refId%22:%22histogram_buckets%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22prometheus%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22sum%20by%28le%29%20%28rate%28http_server_requests_seconds_bucket%7BprojectId%3D%5C%22influencer-model-dt%5C%22%7D%5B$__rate_interval%5D%29%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true,%22hide%22:false%7D%5D,%22range%22:%7B%22from%22:%221697644557037%22,%22to%22:%221697644563234%22%7D%7D&orgId=1)).

### What did you expect to see?

I was expecting to see a value closer to `22.4ms` for the `histogram_quantile(1)`, not my largest configured bucket histogram `le` value (`le=60s`).

This same behaviour occurs for other quantiles, like for `histogram_quantile(0.9999)`, albeit more rarely. It's very important in our case though to monitor very high latency quantiles, like `q0.9999`.

Am I doing something wrong?

Thank you!

### System information

```text
Linux 5.10.186-179.751.amzn2.x86_64 x86_64
```

### Prometheus version

```text
v2.42.0 as distributed here: http://quay.io/prometheus/prometheus:v2.42.0
```

**Comment by bboreham** [comment]
I've seen something like this before, where the counts differed by a really tiny amount like 1E-12.
You are showing `rate()` in the last couple of pictures; you need to show the raw counts, and also be sure that nothing is rounding the values, before this can be discounted.

**Comment by AlbertoEAF** [comment]
Thank you for answering @bboreham 

By raw counts you mean the values of `http_server_requests_seconds_buckets`? 

Before we go to raw counts though, even looking at `y = histogram_quantile(q, D)` where `D` is the data shown, `histogram_quantile` should have computed a very different output `y` with the inputs given as per the screenshot above, right?

**P.S.:** I'm not sure which raw counts to show / what to look for in them -- I have 1650 of them at this instant, and several pods emitting series with the same `le` buckets. And then `rate()` which operates on those also needs to look at prior raw counts as well for at least a time window spanning `$__rate_interval=2m`.

**Comment by bboreham** [comment]
> By raw counts you mean the values of http_server_requests_seconds_buckets? 

Yes, those are counts.

> Or simply the raw outputs of sum by (le) (rate(...)) without rounding? 

Those are rates, not counts.

> Raw CSV data taken from Grafana 

I still wouldn't be certain that is not rounding the data.  Get it raw from Prometheus.

**Comment by beorn7** [comment]
I'm not familiar with what Grafan is doing with the date. I would rather look at the data in the Prometheus UI or `curl` results from the Prometheus endpoint directly when debugging these kind of low-level issues. 

If I read the CSV dump correctly, then the `+Inf` bucket value in the first row is 6012.146074655326, which is actually _less_ than many other values (and that by quite a bit, so that it might be more than just floating point precision to blame). In the 2nd row, the `+Inf` value is 8006.934659115397, which is indeed _greater_ than any other value. So here `histogram_quantile` will correctly calculate the largest bucket boundary as the estimated value for the 100th percentile.

In summary, I think that `histogram_quantile` is doing the right thing, and that the error must be in the data.

**Comment by naveen-ramanathan-form3** [comment]
I am having this same issue when using quantile=1

**Comment by bboreham** [comment]
The status of this issue is that we asked to see the raw output from Prometheus, and cannot comment further without it.

If you have a similar issue please click the 'new issue' button and supply full details.
